File: core/src/com/broll/gainea/server/core/utils/ShipUtils.java
package com.broll.gainea.server.core.utils;

import com.broll.gainea.server.core.map.Area;
import com.broll.gainea.server.core.map.AreaCollection;
import com.broll.gainea.server.core.map.Expansion;
import com.broll.gainea.server.core.map.Island;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.map.Ship;

import java.util.List;
import java.util.stream.Collectors;

public class ShipUtils {

    public static Area targetLocation(Ship ship) {
        Location to;
        do {
            to = ship.getTo();
        } while (to instanceof Ship);
        return (Area) to;
    }

    public static Area sourceLocation(Ship ship) {
        Location from;
        do {
            from = ship.getFrom();
        } while (from instanceof Ship);
        return (Area) from;
    }

    public static boolean leadsTo(Ship ship, Area area) {
        Area to = targetLocation(ship);
        return to.getNumber() == area.getNumber();
    }

    public static boolean leadsTo(Ship ship, AreaCollection areaCollection) {
        Area to = targetLocation(ship);
        return to.getContainer() == areaCollection;
    }

    public static boolean leadsTo(Ship ship, Expansion expansion) {
        Area to = targetLocation(ship);
        return to.getContainer().getExpansion() == expansion;
    }

    public static boolean startsFrom(Ship ship, Area area) {
        Area from = sourceLocation(ship);
        return from.getNumber() == area.getNumber();
    }

    public static boolean startsFrom(Ship ship, AreaCollection areaCollection) {
        Area from = sourceLocation(ship);
        return from.getContainer() == areaCollection;
    }

    public static boolean startsFrom(Ship ship, Expansion expansion) {
        Area from = sourceLocation(ship);
        return from.getContainer().getExpansion() == expansion;
    }

    public static boolean connects(Ship ship, Area from, Area to) {
        return startsFrom(ship, from) && leadsTo(ship, to);
    }

    public static boolean connects(Ship ship, AreaCollection from, AreaCollection to) {
        return startsFrom(ship, from) && leadsTo(ship, to);
    }

    public static boolean connects(Ship ship, Expansion from, Expansion to) {
        return startsFrom(ship, from) && leadsTo(ship, to);
    }

    public static List<Ship> getShips(AreaCollection from, AreaCollection to) {
        return from.getShips().stream().filter(it -> ShipUtils.leadsTo(it, to)).collect(Collectors.toList());
    }
}

File: server/src/main/com/broll/gainea/server/core/utils/ShipUtils.kt
package com.broll.gainea.server.core.utils

import com.broll.gainea.server.core.map.Area
import com.broll.gainea.server.core.map.AreaCollection
import com.broll.gainea.server.core.map.Expansion
import com.broll.gainea.server.core.map.Location
import com.broll.gainea.server.core.map.Ship

fun Ship.targetLocation(): Area {
    var to: Location = this
    do {
        to = (to as Ship).to
    } while (to is Ship)
    return to as Area
}

fun Ship.sourceLocation(): Area {
    var from: Location = this
    do {
        from = (from as Ship).from
    } while (from is Ship)
    return from as Area
}

fun Ship.leadsTo(area: Area) = targetLocation() == area

fun Ship.leadsTo(areaCollection: AreaCollection) = targetLocation().container == areaCollection

fun Ship.leadsTo(expansion: Expansion) = targetLocation().container.expansion == expansion

fun Ship.startsFrom(area: Area) = sourceLocation() == area

fun Ship.startsFrom(areaCollection: AreaCollection) = sourceLocation().container == areaCollection


fun Ship.startsFrom(expansion: Expansion) = sourceLocation().container.expansion == expansion

fun Ship.connects(from: Area, to: Area) = startsFrom(from) && leadsTo(to)


fun Ship.connects(from: AreaCollection, to: AreaCollection) = startsFrom(from) && leadsTo(to)


fun Ship.connects(from: Expansion, to: Expansion) = startsFrom(from) && leadsTo(to)


fun AreaCollection.getAllShips(to: AreaCollection) =
        ships.filter { it.leadsTo(to) }

fun AreaCollection.getAllShips() =
        listOf(ships,
                expansion.contents.filter { it != this }.flatMap { it.getAllShips(this) }
        ).flatten()


